# Imports

In [29]:
import pandas as pd
pd.set_option('display.max_columns', None)
import numpy as np
import matplotlib.pyplot as plt
import json

In [30]:
raw_data = pd.read_csv('hotel_bookings.csv')
raw_data.sample(8)

hotel  is_canceled  lead_time  arrival_date_year  \
16839  Resort Hotel            0         60               2015   
97927    City Hotel            0        249               2016   
49792    City Hotel            1        103               2016   
47529    City Hotel            1         70               2016   
95000    City Hotel            0        145               2016   
79479    City Hotel            0         12               2015   
10585  Resort Hotel            1        210               2017   
26476  Resort Hotel            0          4               2016   

      arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
16839          September                        37                          8   
97927          September                        39                         18   
49792              April                        17                         19   
47529           February                         9                         26   
95000             August                        33                         11   
79479            October                        44                         25   
10585              March                        11                         15   
26476               July                        31                         30   

       stays_in_weekend_nights  stays_in_week_nights  adults  children  \
16839                        2                     5       2       0.0   
97927                        2                     2       2       0.0   
49792                        0                     2       1       0.0   
47529                        0                     2       1       0.0   
95000                        0                     2       2       2.0   
79479                        2                     1       2       0.0   
10585                        1                     4       2       0.0   
26476                        1                     1       2       0.0   

       babies meal country market_segment distribution_channel  \
16839       0   BB     IRL      Online TA                TA/TO   
97927       0   BB     FRA         Direct               Direct   
49792       0   BB     PRT  Offline TA/TO                TA/TO   
47529       0   BB     NLD      Online TA                TA/TO   
95000       0   BB     GBR      Online TA                TA/TO   
79479       0   BB     PRT         Groups                TA/TO   
10585       0   BB     NLD         Direct               Direct   
26476       0   HB     ESP      Online TA                TA/TO   

       is_repeated_guest  previous_cancellations  \
16839                  0                       0   
97927                  0                       0   
49792                  0                       0   
47529                  0                       0   
95000                  0                       0   
79479                  0                       0   
10585                  0                       0   
26476                  0                       0   

       previous_bookings_not_canceled reserved_room_type assigned_room_type  \
16839                               0                  D                  E   
97927                               0                  A                  A   
49792                               0                  A                  A   
47529                               0                  A                  A   
95000                               0                  F                  F   
79479                               0                  A                  A   
10585                               0                  E                  F   
26476                               0                  A                  C   

       booking_changes deposit_type  agent  company  days_in_waiting_list  \
16839                0   No Deposit  240.0      NaN                     0   
97927                0   No Deposit   14.0      NaN                     0   
49792          

## 1. Data Collection


In [31]:
raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

- There are 32 columns in the dataset.
- The dataset contains 119,390 rows.
- Data Types:
    - 1. 16 columns are of int64 data type.
    - 2. 4 columns are of float64 data type.
    - 3. 12 columns are of object data types. 
- Missing Values:
    - The column `children` has <span style="color:red">4</span> missing values.
    - The column `country` has <span style="color:red">488</span> missing values.
    - The column `agent` has <span style="color:red">16,340</span> missing values.
    - The column `company` has <span style="color:red">112,593</span> missing values.


## 2 Exploratory Data Analysis (EDA)

### 2.1 Looking for missing values.

In [32]:
raw_data.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

- Dropping null values is easy but isn't a good practice, so we will start with replacing them with zero.

In [33]:
raw_data.fillna(0, inplace=True)

### 2.2 cleaning Data

In [34]:
raw_data.describe().T

count         mean         std      min  \
is_canceled                     119390.0     0.370416    0.482918     0.00   
lead_time                       119390.0   104.011416  106.863097     0.00   
arrival_date_year               119390.0  2016.156554    0.707476  2015.00   
arrival_date_week_number        119390.0    27.165173   13.605138     1.00   
arrival_date_day_of_month       119390.0    15.798241    8.780829     1.00   
stays_in_weekend_nights         119390.0     0.927599    0.998613     0.00   
stays_in_week_nights            119390.0     2.500302    1.908286     0.00   
adults                          119390.0     1.856403    0.579261     0.00   
children                        119390.0     0.103886    0.398555     0.00   
babies                          119390.0     0.007949    0.097436     0.00   
is_repeated_guest               119390.0     0.031912    0.175767     0.00   
previous_cancellations          119390.0     0.087118    0.844336     0.00   
previous_bookings_not_canceled  119390.0     0.137097    1.497437     0.00   
booking_changes                 119390.0     0.221124    0.652306     0.00   
agent                           119390.0    74.828319  107.141953     0.00   
company                         119390.0    10.775157   53.943884     0.00   
days_in_waiting_list            119390.0     2.321149   17.594721     0.00   
adr                             119390.0   101.831122   50.535790    -6.38   
required_car_parking_spaces     119390.0     0.062518    0.245291     0.00   
total_of_special_requests       119390.0     0.571363    0.792798     0.00   

                                    25%       50%     75%     max  
is_canceled                        0.00     0.000     1.0     1.0  
lead_time                         18.00    69.000   160.0   737.0  
arrival_date_year               2016.00  2016.000  2017.0  2017.0  
arrival_date_week_number          16.00    28.000    38.0    53.0  
arrival_date_day_of_month          8.00    16.000    23.0    31.0  
stays_in_weekend_nights            0.00     1.000     2.0    19.0  
stays_in_week_nights               1.00     2.000     3.0    50.0  
adults                             2.00     2.000     2.0    55.0  
children                           0.00     0.000     0.0    10.0  
babies                             0.00     0.000     0.0    10.0  
is_repeated_guest                  0.00     0.000     0.0     1.0  
previous_cancellations             0.00     0.000     0.0    26.0  
previous_bookings_not_canceled     0.00     0.000     0.0    72.0  
booking_changes                    0.00     0.000     0.0    21.0  
agent                              7.00     9.000   152.0   535.0  
company                            0.00     0.000     0.0   543.0  
days_in_waiting_list               0.00     0.000     0.0   391.0  
adr                               69.29    94.575   126.0  5400.0  
required_car_parking_spaces        0.00     0.000     0.0     8.0  
total_of_special_requests          0.00     0.000     1.0     5.0

In [35]:
# Display rows with negative adr values
negative_adr_rows = raw_data[raw_data['adr'] < 0]
display(negative_adr_rows)

# Drop rows with negative adr values
raw_data.drop(negative_adr_rows.index, inplace=True)

hotel  is_canceled  lead_time  arrival_date_year  \
14969  Resort Hotel            0        195               2017   

      arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
14969              March                        10                          5   

       stays_in_weekend_nights  stays_in_week_nights  adults  children  \
14969                        4                     6       2       0.0   

       babies meal country market_segment distribution_channel  \
14969       0   BB     GBR         Groups               Direct   

       is_repeated_guest  previous_cancellations  \
14969                  1                       0   

       previous_bookings_not_canceled reserved_room_type assigned_room_type  \
14969                               2                  A                  H   

       booking_changes deposit_type  agent  company  days_in_waiting_list  \
14969                2   No Deposit  273.0      0.0                     0   

         customer_type   adr  required_car_parking_spaces  \
14969  Transient-Party -6.38                            0   

       total_of_special_requests reservation_status reservation_status_date  
14969                          0          Check-Out                15-03-17

### Based on the following logic:

1.  Rows containing negative values in the `adr` column.
2.  The `distribution_channel` column, as it depicts the same information as the `market_segment` column.
3.  The `arrival_date_week_number` column, as the week number can be derived from the `arrival_date` column.
4.  Rows where both the `children` and `adults` columns contain zero.

In [36]:
def clean_dataframe(data):
    """
    Cleans a pandas DataFrame based on specified criteria.

    Args:
        raw_data: The pandas DataFrame to clean.

    Returns:
        The cleaned pandas DataFrame.
    """

    # 1. Drop rows with negative 'adr' values.
    data.drop(data[data['adr'] <= 0].index, inplace=True)

    # 2. Drop the 'distribution_channel' column.
    if 'distribution_channel' in data.columns: # check if column exists
        data = data.drop(columns=['distribution_channel'])

    # 3. Drop the 'arrival_date_week_number' column.
    if 'arrival_date_week_number' in data.columns: # check if column exists
        data = data.drop(columns=['arrival_date_week_number'])

    # 4. Drop rows where both 'children' and 'adults' are zero.
    data = data[~((data['children'] == 0) & (data['adults'] == 0))] # ~ is the not operator.

    return data


raw_data = clean_dataframe(raw_data)

In [37]:
raw_data

hotel  is_canceled  lead_time  arrival_date_year  \
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
5       Resort Hotel            0         14               2015   
6       Resort Hotel            0          0               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_day_of_month  stays_in_weekend_nights  \
2                    July                          1                        0   
3                    July                          1                        0   
4                    July                          1                        0   
5                    July                          1                        0   
6                    July                          1                        0   
...                   ...                        ...                      ...   
119385             August                         30                        2   
119386             August                         31                        2   
119387             August                         31                        2   
119388             August                         31                        2   
119389             August                         29                        2   

        stays_in_week_nights  adults  children  babies meal country  \
2                          1       1       0.0       0   BB     GBR   
3                          1       1       0.0       0   BB     GBR   
4                          2       2       0.0       0   BB     GBR   
5                          2       2       0.0       0   BB     GBR   
6                          2       2       0.0       0   BB     PRT   
...                      ...     ...       ...     ...  ...     ...   
119385                     5       2       0.0       0   BB     BEL   
119386                     5       3       0.0       0   BB     FRA   
119387                     5       2       0.0       0   BB     DEU   
119388                     5       2       0.0       0   BB     GBR   
119389                     7       2       0.0       0   HB     DEU   

       market_segment  is_repeated_guest  previous_cancellations  \
2              Direct                  0                       0   
3           Corporate                  0                       0   
4           Online TA                  0                       0   
5           Online TA                  0                       0   
6              Direct                  0                       0   
...               ...                ...                     ...   
119385  Offline TA/TO                  0                       0   
119386      Online TA                  0                       0   
119387      Online TA                  0                       0   
119388      Online TA                  0                       0   
119389      Online TA                  0                       0   

        previous_bookings_not_canceled reserved_room_type assigned_room_type  \
2                                    0                  A                  C   
3                                    0                  A                  A   
4                                    0                  A                  A   
5                                    0                  A                  A   
6                                    0                  C                  C   
...                                ...                ...                ...   
119385 

### 2.3 Feature Selection

##### Why Feature Selection Might Be Beneficial:

* **Improved LLM Performance:** LLMs can be sensitive to irrelevant or redundant information. Feature selection can help focus the LLM on the most important data points, potentially improving accuracy and relevance.

* **Reduced Dimensionality:** If your dataset has a large number of features, feature selection can reduce the dimensionality of the data. This can lead to:

    1.  Faster vector embedding generation and retrieval.
    2.  Reduced storage requirements for your vector database.
    3.  Improved efficiency of your API.

In [38]:
raw_data['booking_source'] = np.where(
    (raw_data['agent'] > 0) & (raw_data['agent'].notna()), 
    'agent',
    np.where(
        (raw_data['company'] > 0) & (raw_data['company'].notna()),
        'company',
        'self'
    )
)

# Enhanced customer history features (FIXED VERSION)
raw_data['previous_experience'] = np.where(
    raw_data['is_repeated_guest'] == 1,
    'Repeat',
    'First-time'
)

# For previous cancellations analysis
raw_data['previous_cancel_behavior'] = np.where(
    raw_data['previous_cancellations'] > 0,
    'Previous Canceller',
    'No Cancel History'
)

# Create customer segments
conditions = [
    raw_data['previous_cancellations'] > 0,
    raw_data['is_repeated_guest'] == 1,
    raw_data['total_of_special_requests'] > 2
]
choices = ['Risk', 'Loyal', 'High-Maintenance']
raw_data['customer_segment'] = np.select(conditions, choices, default='Standard')

raw_data['arrival_date'] = pd.to_datetime(
    raw_data['arrival_date_year'].astype(str) + '-' +
    raw_data['arrival_date_month'] + '-' +
    raw_data['arrival_date_day_of_month'].astype(str)
)

columns_to_drop = [
    'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month',
    'assigned_room_type','days_in_waiting_list', 'reservation_status_date',
    'agent', 'company','is_repeated_guest', 'previous_cancellations','reservation_status'
]
raw_data.drop(columns=columns_to_drop, inplace=True)

- Combinied the `agent` and `company` columns into a single `booking_source` column.
- Determines whether a booking was made through an **agent**, a **company**, or directly by the customer (**self**).

- Created a new column `previous_experience` to categorise customers as either **Repeat** or **First-time** guests.
- Created a new column `previous_cancel_behavior` to categorise customers based on their cancellation history.
- Combined `arrival_date_year`, `arrival_date_month`, and `arrival_date_day_of_month` into a single `arrival_date` column.
- Removed redundant or unnecessary columns from the dataset, to focuses on features that are most relevant for analytics and chatbot queries.


In [39]:
raw_data.to_csv('cleaned_hotel_bookings.csv', index=False)

### 2.4 Chatbot-Focused Text Representation

In [40]:
def analyze_data(df):
    # Convert date column
    df['arrival_date'] = pd.to_datetime(df['arrival_date'])
    
    # Calculate derived columns
    df['total_nights'] = df['stays_in_weekend_nights'] + df['stays_in_week_nights']
    df['traveler_type'] = np.where(
        (df['children'] > 0) | (df['babies'] > 0), 'family', 'solo/couple'
    )

    analysis = {
        # Basic Metrics
        'total_bookings': len(df),
        'cancellation_rate': round(df['is_canceled'].mean() * 100, 2),
        
        # Temporal Analysis (1)
        'lead_time_analysis': {
            'average_lead_time': round(df['lead_time'].mean(), 1),
            'lead_time_cancellation_corr': round(df['lead_time'].corr(df['is_canceled']), 2),
            'lead_time_groups': pd.cut(df['lead_time'], bins=[0,7,30,90,365], 
                                    labels=['0-7', '8-30', '31-90', '90+']).value_counts().to_dict()
        },
        
        # Stay Patterns (2,12)
        'stay_analysis': {
            'average_total_nights': round(df['total_nights'].mean(), 1),
            'weekend_week_ratio': round(
                df['stays_in_weekend_nights'].sum() / df['stays_in_week_nights'].sum(), 2),
            'stay_duration_impact': df.groupby('total_nights').agg({
                'adr': 'mean',
                'is_canceled': 'mean'
            }).rename(columns={'adr':'avg_adr', 'is_canceled':'cancellation_rate'}).to_dict()
        },
        
        # Customer Value (3)
        'customer_analysis': {
            'repeat_customers': df[df['previous_bookings_not_canceled'] > 0].shape[0],
            'customer_type_stats': df.groupby('customer_type').agg({
                'adr': ['mean', 'count'],
                'is_canceled': 'mean'
            }).to_dict()
        },
        
        # Market & Channel (4,10)
        'market_analysis': {
            'segment_profitability': df.groupby('market_segment').agg({
                'adr': 'mean',
                'is_canceled': 'mean'
            }).rename(columns={'adr':'avg_adr'}).to_dict(),
            'channel_performance': df.groupby('booking_source').agg({
                'adr': 'sum',
                'lead_time': 'mean',
                'is_canceled': 'mean'
            }).rename(columns={
                'adr':'total_revenue',
                'lead_time':'avg_lead_time'
            }).to_dict()
        },
        
        # Behavioral Factors (5,6)
        'behavior_analysis': {
            'deposit_impact': df.groupby('deposit_type')['is_canceled'].mean().to_dict(),
            'special_requests_impact': df.groupby('total_of_special_requests').agg({
                'adr': 'mean',
                'is_canceled': 'mean'
            }).rename(columns={'adr':'avg_adr'}).to_dict()
        },
        
        # Existing Metrics (kept for compatibility)
        'room_metrics': {
            'most_popular_room': df['reserved_room_type'].mode()[0],
            'type_distribution': df['reserved_room_type'].value_counts().to_dict()
        },
        'revenue_analysis': {
            'monthly_revenue': {k.strftime('%Y-%m'): v for k,v in 
                               df.groupby(df['arrival_date'].dt.to_period('M'))['adr'].sum().items()},
            'peak_month': df.groupby(df['arrival_date'].dt.to_period('M'))['adr'].sum().idxmax().strftime('%Y-%m')
        },
        'guest_demographics': {
            'top_countries': df['country'].value_counts().head(5).to_dict(),
            'meal_plan_distribution': df['meal'].value_counts().to_dict()
        }
    }
    
    # Convert numpy types to native Python types
    for category in analysis:
        if isinstance(analysis[category], dict):
            for key in analysis[category]:
                if isinstance(analysis[category][key], (np.generic, pd.Timestamp)):
                    analysis[category][key] = analysis[category][key].item()
    
    return analysis

**What’s Done:**
- **Date Conversion:** Converts the `arrival_date` column to a datetime format for temporal analysis.
- **Derived Columns:**
  - `total_nights`: Sums `stays_in_weekend_nights` and `stays_in_week_nights` to calculate total stay duration.
  - `traveler_type`: Classifies guests as "family" (if they have children/babies) or "solo/couple".

### **1. Basic Metrics**
**What’s Done:**
- **Total Bookings:** Counts the number of records (bookings).
- **Cancellation Rate:** Calculates the percentage of bookings marked as canceled (`is_canceled`).

**Why It’s Done:**
- **Baseline Understanding:** Provides the scale of operations (total bookings) and a critical performance indicator (cancellation rate), which impacts revenue stability.


### **2. Temporal Analysis**
**What’s Done:**
- **Average Lead Time:** Measures the average days between booking and arrival.
- **Lead Time vs. Cancellation Correlation:** Checks if longer lead times correlate with higher cancellation likelihood.
- **Lead Time Groups:** Bins lead times into categories (e.g., 0-7 days, 8-30 days) and calculates their distribution.

### **3. Stay Patterns**
**What’s Done:**
- **Average Total Nights:** Measures the typical stay duration.
- **Weekend/Week Ratio:** Compares weekend vs. weekday stays.
- **Stay Duration Impact:** Analyzes how stay length affects average daily rate (ADR) and cancellation rates.

**Why It’s Done:**
- **Resource Allocation:** Helps optimize staffing/room inventory based on typical stay lengths.
- **Pricing Strategy:** Identifies if longer stays correlate with higher/lower ADR or cancellations.

### **4. Customer Value**
**What’s Done:**
- **Repeat Customers:** Counts guests with prior non-canceled bookings.
- **Customer Type Stats:** Groups by `customer_type` (e.g., transient, group) to analyze ADR, booking count, and cancellation rates.


### **5. Market & Channel Analysis**
**What’s Done:**
- **Market Segment Profitability:** Groups by `market_segment` (e.g., corporate, online) to compare ADR and cancellations.
- **Channel Performance:** Groups by `booking_source` to analyze revenue, lead time, and cancellations.

### **6. Behavioral Factors**
**What’s Done:**
- **Deposit Impact:** Analyzes cancellation rates by deposit type (e.g., non-refundable vs. refundable).
- **Special Requests:** Groups by `total_of_special_requests` to see how requests affect ADR and cancellations.


### **7. Existing Metrics (for Compatibility)**
**What’s Done:**
- **Room Popularity:** Identifies the most booked room type.
- **Monthly Revenue:** Aggregates revenue by month and identifies peak months.
- **Guest Demographics:** Lists top countries and meal preferences.


### **8. Data Type Conversion**
**What’s Done:**
- Converts numpy/pandas-specific types (e.g., `np.int64`, `pd.Timestamp`) to native Python types (e.g., `int`, `str`).

### **Key Business Questions Addressed**
This analysis helps answer questions like:
- Why do cancellations happen (lead time, deposit policies)?
- Which customer segments are most profitable?
- How do booking channels differ in performance?
- When is demand highest (seasonality)?
- What drives guest loyalty (repeat bookings)?

By structuring insights into a nested dictionary, the code prepares data for LLMs to generate human-readable reports, recommendations, or dashboards.

In [47]:
# In your Jupyter notebook
import json
import numpy as np
from pandas import Period

def prepare_rag_data(analysis, filename="formatted_analysis.json"):
    """Convert nested analysis to RAG-friendly format and save locally"""
    
    # 1. Type Conversion Helper
    def convert_value(value):
        if isinstance(value, (np.generic)):
            return float(value) if 'float' in str(value.dtype) else int(value)
        if isinstance(value, Period):
            return value.strftime('%Y-%m')
        if isinstance(value, dict):
            return {k: convert_value(v) for k, v in value.items()}
        return value
    
    # 2. Flatten Analysis Structure
    def flatten_analysis(data, parent_key='', sep='_'):
        items = []
        for k, v in data.items():
            new_key = f"{parent_key}{sep}{k}" if parent_key else k
            if isinstance(v, dict):
                items.extend(flatten_analysis(v, new_key, sep=sep).items())
            else:
                items.append((new_key, convert_value(v)))
        return dict(items)
    
    # 3. Create Document Chunks
    formatted_data = []
    flat_analysis = flatten_analysis(analysis)
    
    for key, value in flat_analysis.items():
        # Create human-readable documents
        doc = {
            "text": f"{key.replace('_', ' ').title()}: {value}",
            "metadata": {
                "category": key.split('_')[0],
                "data_type": str(type(value)).split("'")[1]
            }
        }
        formatted_data.append(doc)
    
    # 4. Save to Local File
    with open(filename, 'w') as f:
        json.dump(formatted_data, f, indent=2)
    
    print(f"Saved {len(formatted_data)} RAG-ready documents to {filename}")
    return formatted_data

# Usage in Notebook
analysis = analyze_data(raw_data)  # Your existing analysis function
rag_ready_data = prepare_rag_data(analysis)

Saved 188 RAG-ready documents to formatted_analysis.json


In [6]:
import gc, torch
torch.cuda.empty_cache()
gc.collect()

9025